### PR TITLE
Fine tunning implementations [12984]

### PR DESCRIPTION
**Pull request by jparisu**
Adding comments, logs, delete functions, to Participants, Writers, Readers and ParticipantFactory

Merge after #52

### CODE REVIEW

File: test/blackbox/ddsrouter_core/TrivialTest.cpp [hunk 1]
```diff
@@ -36,81 +36,87 @@ Guid random_guid(
...
-//     data.payload = payload;
+    DummyDataReceived data;
+    data.source_guid = guid;
+    data.payload = payload;
 
-//     participant_1->add_message_to_send(topic, data);
+    participant_1->simulate_data_reception(topic, data);
 
-//     std::this_thread::sleep_for(std::chrono::seconds(2));
+    // Sleep until first data arrive to the writer
```
Comment (juanlofer-eprosima) [inline test/blackbox/ddsrouter_core/TrivialTest.cpp:109]:
```suggestion
    // Sleep until first data arrives to the writer
```

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 2]
```diff
@@ -175,3 +175,21 @@ if(BUILD_DOCUMENTATION OR BUILD_DOCUMENTATION_TESTS)
...
+###############################################################################
+# In Debug, log set to ON
+if ("Debug" STREQUAL ${CMAKE_BUILD_TYPE})
+    option(LOG_INFO "Compile logInfo messages" ON)
+else()
+    option(LOG_INFO "Compile logInfo messages" OFF)
+endif()
+
+if (LOG_INFO)
+    message(STATUS "Compiling logInfo messages for Databroker")
```
[the remark this answers is not in the document]
Comment (jparisu) [reply CMakeLists.txt:190]: ups :+1:

File: include/ddsrouter/participant/implementations/auxiliar/impl/BaseParticipant.ipp [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    : configuration_(participant_configuration)
+    , payload_pool_(payload_pool)
+    , discovery_database_(discovery_database)
+{
+}
+
+template <class ConfigurationType>
+BaseParticipant<ConfigurationType>::~BaseParticipant()
+{
+    std::lock_guard<std::recursive_mutex> lock(mutex_);
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/participant/implementations/auxiliar/impl/BaseParticipant.ipp:47]: :+1:

File: include/ddsrouter/participant/implementations/auxiliar/impl/BaseParticipant.ipp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+    if (!writers_.empty())
+    {
+        logWarning(DDSROUTER_BASEPARTICIPANT, "Deleting Participant " << id() << " with writers");
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/participant/implementations/auxiliar/impl/BaseParticipant.ipp:51]: Be careful, you got it wrong. The idea is to warn that the Participant that is being deleted still has Writers.

File: include/ddsrouter/participant/implementations/auxiliar/impl/BaseParticipant.ipp [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        writers_.clear();
+    }
+
+    if (!readers_.empty())
+    {
+        logWarning(DDSROUTER_BASEPARTICIPANT, "Deleting Participant " << id() << " with readers");
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/participant/implementations/auxiliar/impl/BaseParticipant.ipp:57]: Same as before

File: src/cpp/writer/implementations/auxiliar/BaseWriter.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+namespace eprosima {
+namespace ddsrouter {
+
+BaseWriter::BaseWriter(
+        const ParticipantId& participant_id,
+        const RealTopic& topic,
+        std::shared_ptr<PayloadPool> payload_pool)
+    : participant_id_(participant_id)
+    , topic_(topic)
+    , payload_pool_(payload_pool)
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/writer/implementations/auxiliar/BaseWriter.cpp:32]: :+1: (good catch)

File: include/ddsrouter/reader/IReader.hpp [hunk 7]
```diff
@@ -30,29 +30,74 @@ namespace eprosima {
...
+     * After this method, the Reader should not notify any message that arrives.
+     */
+    virtual void unset_on_data_available_callback() noexcept = 0;
 
+    /**
+     * @brief Take oldest received message from the Reader
+     *
+     * This method will take the oldest sample received by the Reader and will set it to the argument \c data
+     * in a way that the payload in \c data must be inside the DDS Router PayloadPool.
+     * In \c data must be specified the Guid of the Endpoint that originally send this data.
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/reader/IReader.hpp:90]: :+1:

File: include/ddsrouter/reader/implementations/auxiliar/DummyReader.hpp [hunk 8]
```diff
@@ -23,57 +23,59 @@
...
-
-    ReturnCode take(
-            std::unique_ptr<DataReceived>& data_received) override;
+    //! Use parent constructors
+    using BaseReader::BaseReader;
 
-    void add_message_to_send(
-            DataToSend data);
+    /**
+     * @brief Make the Reader to simulate data reception
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/reader/implementations/auxiliar/DummyReader.hpp:53]: :+1:

File: include/ddsrouter/writer/IWriter.hpp [hunk 9]
```diff
@@ -28,18 +28,56 @@ namespace eprosima {
...
+     * By default the Writer is disable. Call this method to activate it.
+     */
+    virtual void enable() noexcept = 0;
 
-    virtual void disable() = 0;
+    /**
+     * @brief Disable Writer
+     *
+     * A Writer disabled do not send data.
+     * @note Method \c write from a disabled writer should never be called
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/writer/IWriter.hpp:57]: :+1:

File: include/ddsrouter/writer/implementations/auxiliar/BaseWriter.hpp [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+#include <ddsrouter/writer/IWriter.hpp>
+#include <ddsrouter/types/participant/ParticipantId.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+
+/**
+ * Writer that has an empty implementation.
+ * It does not send anything.
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/writer/implementations/auxiliar/BaseWriter.hpp:33]: :+1:

File: include/ddsrouter/writer/implementations/auxiliar/BaseWriter.hpp [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+ */
+class BaseWriter : public IWriter
+{
+public:
+
+    /**
+     * @brief Construct a new Base Writer object
+     *
+     * @param participant_id parent participant id
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/writer/implementations/auxiliar/BaseWriter.hpp:42]: :3rd_place_medal:

File: include/ddsrouter/writer/implementations/auxiliar/DummyWriter.hpp [hunk 12]
```diff
@@ -19,44 +19,80 @@
...
 #include <ddsrouter/types/Timestamp.hpp>
-#include <ddsrouter/writer/implementations/auxiliar/EchoWriter.hpp>
+#include <ddsrouter/writer/implementations/auxiliar/BaseWriter.hpp>
 #include <ddsrouter/writer/IWriter.hpp>
 
 namespace eprosima {
 namespace ddsrouter {
 
-struct DataStored
+//! Data that a Dummy Writer should have been sent
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/writer/implementations/auxiliar/DummyWriter.hpp:33]: :+1:

File: src/cpp/writer/implementations/auxiliar/BaseWriter.cpp [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+{
+    std::lock_guard<std::recursive_mutex> lock(mutex_);
+
+    if (enabled_.load())
+    {
+        return write_(data);
+    }
+    else
+    {
+        logWarning(DDSROUTER_BASEREADER, "Attempt to take data from disabled Reader in topic " <<
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/writer/implementations/auxiliar/BaseWriter.cpp:76]: :+1:

File: include/ddsrouter/writer/implementations/auxiliar/BaseWriter.hpp [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+#include <mutex>
+
+#include <ddsrouter/types/participant/ParticipantId.hpp>
+#include <ddsrouter/writer/IWriter.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+
+/**
+ * Abstract Writer that implement generic methods for every Writer.
```
Comment (juanlofer-eprosima) [inline include/ddsrouter/writer/implementations/auxiliar/BaseWriter.hpp:32]:
```suggestion
 * Abstract Writer that implements generic methods for every Writer.
```

File: include/ddsrouter/writer/implementations/auxiliar/BaseWriter.hpp [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * In order to inherit this class, create the protected method write_ .
```
Comment (juanlofer-eprosima) [inline include/ddsrouter/writer/implementations/auxiliar/BaseWriter.hpp:34]:
```suggestion
 * In order to inherit from this class, create the protected method write_ .
```

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Pull request merged**